pkg/route/validation/validation.go

```diff
@@ -342,6 +370,30 @@ func validateWildcardPolicy(host string, policy routev1.WildcardPolicyType, fldP
 	return nil
 }
 
+var (
+	notAllowedHTTPHeaderResponse    = []string{"strict-transport-security"}
+	notAllowedHTTPHeaderResponseSet = sets.NewString(notAllowedHTTPHeaderResponse...)
+)
+
+func validateHeaders(path *field.Path, headerValues []routev1.RouteHTTPHeader) *field.Error {
+	for key, value := range headerValues {
+		if value.Action.Set != nil && len(value.Action.Set.Value) != 0 && len(value.Action.Set.Value) > 4096 {
+			return field.Invalid(path.Index(key).Child("action.set.value"), value.Name, fmt.Sprintf("value exceeds the maximum length which is 2048"))
+		}
+		if value.Action.Set != nil && len(value.Action.Set.Value) == 0 {
+			return field.Invalid(path.Index(key).Child("action.set.value"), value.Name, fmt.Sprintf("minimum length of value shall be one."))
+		}
+		if notAllowedHTTPHeaderResponseSet.Has(strings.ToLower(value.Name)) {
+			return field.NotSupported(path.Index(key).Child("name"), value.Name, []string{fmt.Sprintf("All headers except %v", strings.Join(notAllowedHTTPHeaderResponse, ","))})
+		}
+		r, _ := regexp.Compile("^(?:%(?:%|(?:\\{[-+]?[QXE](,[-+]?[QXE])*\\})?\\[(?:req\\.hdr\\([0-9A-Za-z-]+\\)|ssl_c_der)(?:,(?:lower|base64))*\\])|[^%[:cntrl:]])*$")
```

Comment:
I'd like to see more extensive unit testing for the set of permissible header values this regexp supports. I think it would be useful to write a discrete test for this regexp as I did here: https://github.com/frobware/haproxy-hacks/blob/master/headers/header_test.go. My tests are not exhaustive. I wanted to understand the failure cases, for which (assuming my tests are sane) I get the following unit test failures:

```console
% go test
--- FAIL: TestSamples (0.00s)
    --- FAIL: TestSamples/empty_string (0.00s)
        header_test.go:54: "": expected false, got true
    --- FAIL: TestSamples/multiple_words_without_escaped_space (0.00s)
        header_test.go:54: "abc def": expected false, got true
    --- FAIL: TestSamples/escaped_expression (0.00s)
        header_test.go:54: "%%[src]\\ %[src]": expected true, got false
    --- FAIL: TestSamples/request_hdr_with_percent_field (0.00s)
        header_test.go:54: "%[req.hdr(%)]": expected true, got false
    --- FAIL: TestSamples/quoted_field (0.00s)
        header_test.go:54: "%[req.hdr(%{+Q}Host)]": expected true, got false
    --- FAIL: TestSamples/value_with_conditional_expression (0.00s)
        header_test.go:54: "%[req.hdr(Host)] if foo": expected false, got true
FAIL
exit status 1
FAIL    github.com/frobware/haproxy-hacks/headers       0.002s

```

Anywhere this regexp permits an erroneous value to end up in the haproxy.cofng will break routing. For example, if you supply an empty value for a header (which the regex permits):

```config
backend default
  log global
  mode http
  server s1 :9090


  # "empty string"
  http-request set-header Testhdr_0 
```

```console
% haproxy-2.6.6 -f ./haproxy.cfg  -d -V
[NOTICE]   (139315) : haproxy version is 2.6.6-274d1a4
[NOTICE]   (139315) : path to executable is /etc/profiles/per-user/aim/bin/haproxy-2.6.6
[ALERT]    (139315) : config : parsing [./haproxy.cfg:28] : error detected in backend 'default' while parsing 'http-request set-header' rule : expects exactly 2 arguments.
[ALERT]    (139315) : config : Error(s) found in configuration file : ./haproxy.cfg
```

Given the regex approach I'm concerned that unless we've covered all the bases it's very easy to break routing (for the cluster).

Sample `haproxy.cfg` with some of my experiments: https://github.com/frobware/haproxy-hacks/blob/master/headers/haproxy.cfg.

Reply:
@frobware This format -> ^(%(%|(\\{[-+]?[QXE](,[-+]?[QXE])*\\})?([A-Za-z]+|\\[[.0-9A-Z_a-z]+(\\([^)]+\\))?(,[.0-9A-Z_a-z]+(\\([^)]+\\))?)*\\]))|[^%[:cntrl:]])*$    I think will pass the use cases you provided.
However it will not check for the allowed converters. Shall we then just check the format of the header value using the above regex to avoid haproxy to fail rather than checking allowed list of converters ? cc @Miciah

Reply:
@frobware we do check for empty value https://github.com/openshift/library-go/pull/1449/commits/e12b78a1cf2c964ccbfb3c9948e4022a8e4f1a72#diff-bfc6ba95cbc94986901ce1d0e3b74ee966499eab85a96e1d1ff763d2d74a7eeeR383

Reply:
--- FAIL: TestSamples/empty_string (0.00s)
        header_test.go:54: "": expected false, got true    --- we check it as mentioned above
    --- FAIL: TestSamples/multiple_words_without_escaped_space (0.00s)
        header_test.go:54: "abc def": expected false, got true  ---- is expected and correct, got true is correct
    --- FAIL: TestSamples/escaped_expression (0.00s)
        header_test.go:54: "%%[src]\\ %[src]": expected true, got false  ---- is expected and got false is correct
    --- FAIL: TestSamples/request_hdr_with_percent_field (0.00s)
        header_test.go:54: "%[req.hdr(%)]": expected true, got false ----is expected and got false is correct
    --- FAIL: TestSamples/quoted_field (0.00s)
        header_test.go:54: "%[req.hdr(%{+Q}Host)]": expected true, got false   -- is expected and got false is correct
    --- FAIL: TestSamples/value_with_conditional_expression (0.00s)
        header_test.go:54: "%[req.hdr(Host)] if foo": expected false, got true --- is expected ad got true is correct

Reply:
> we do check for empty value

I see; good to know.

There's still the other cases, having explored this a little, that continue to be a bother:

```
set-header %[req.hdr(Host)] if foo
```

should we support conditionals? The regex currently support this. Even if we do sypport conditionals, we don't have enough context to know whether `foo` is actually a variable. If `foo` doesn't exist, then haproxy will fail to start.

Reply:
The regex seems to be doing it's job correctly as mentioned above ^^

Reply:
@miheer Why is the following header value not valid? `"%%[src]\\ %[src]"`?

```console
    --- FAIL: TestSamples/escaped_expression (0.00s)
        header_test.go:55: "%%[src]\\ %[src]": expected true, got false
```

If I plumb this into haproxy and run my test I see:

```
2023/02/23 16:19:16 Testhdr_12: ["%[src] 127.0.0.1"]
```

which is what I intended.

Reply:
This case: 

```console
--- FAIL: TestSamples/request_hdr_with_percent_field (0.00s)
    header_test.go:54: "%[req.hdr(%)]": expected true, got false ----is expected and got false is correct
```

we discussed on Slack[1]. "%" is valid here.

[1] https://redhat-internal.slack.com/archives/CBWMXQJKD/p1677167468811409?thread_ts=1677166429.635719&cid=CBWMXQJKD

Reply:
```console
--- FAIL: TestSamples/empty_string (0.00s)
    header_test.go:54: "": expected false, got true    --- we check it as mentioned above

```

Should the regex require at least one character (assuming no interpolation using %[])?

Reply:
> @miheer Why is the following header value not valid? `"%%[src]\\ %[src]"`?
> 
> ```
>     --- FAIL: TestSamples/escaped_expression (0.00s)
>         header_test.go:55: "%%[src]\\ %[src]": expected true, got false
> ```
> 
> If I plumb this into haproxy and run my test I see:
> 
> ```
> 2023/02/23 16:19:16 Testhdr_12: ["%[src] 127.0.0.1"]
> ```
> 
> which is what I intended.

src is not allowed

Reply:
> ```
> --- FAIL: TestSamples/empty_string (0.00s)
>     header_test.go:54: "": expected false, got true    --- we check it as mentioned above
> ```
> 
> Should the regex require at least one character (assuming no interpolation using %[])?

We are already checking for min length of 1. not sure if we need validation to check one char using regex. @Miciah WDYT ?